lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+	returnThreshold := int(math.Ceil(float64(dht.bucketSize) * OptProvReturnRatio))
+
+	return &estimatorState{
+		putCtx:              ctx,
+		dht:                 dht,
+		key:                 key,
+		doneChan:            make(chan struct{}, returnThreshold), // buffered channel to not miss events
+		ksKey:               ks.XORKeySpace.Key([]byte(key)),
+		networkSize:         networkSize,
+		peerStates:          map[peer.ID]addProviderRPCState{},
+		individualThreshold: individualThreshold,
+		setThreshold:        setThreshold,
+		returnThreshold:     returnThreshold,
+		putProvDone:         atomic.Int32{},
+	}, nil
+}
+
+func (dht *IpfsDHT) GetAndProvideToClosestPeers(outerCtx context.Context, key string) error {
+	if key == "" {
+		return fmt.Errorf("can't lookup empty key")
+	}
+
+	// initialize new context for all putProvider operations.
+	// We don't want to give the outer context to the put operations as we return early before all
+	// put operations have finished to avoid the long tail of the latency distribution. If we
+	// provided the outer context the put operations may be cancelled depending on what happens
+	// with the context on the user side.
+	putCtx, putCtxCancel := context.WithTimeout(context.Background(), time.Minute)
+
+	es, err := dht.newEstimatorState(putCtx, key)
+	if err != nil {
+		putCtxCancel()
+		return err
+	}
+
+	// initialize context that finishes when this function returns
+	innerCtx, innerCtxCancel := context.WithCancel(outerCtx)
+	defer innerCtxCancel()
+
+	go func() {
+		select {
+		case <-outerCtx.Done():
+			// If the outer context gets cancelled while we're still in this function. We stop all
+			// pending put operations.
+			putCtxCancel()
+		case <-innerCtx.Done():
+			// We have returned from this function. Ignore cancellations of the outer context and continue
+			// with the remaining put operations.
+		}
+	}()
+
+	lookupRes, err := dht.runLookupWithFollowup(outerCtx, key, dht.pmGetClosestPeers(key), es.stopFn)
+	if err != nil {
+		return err
+	}
+
+	// Store the provider records with all of the closest peers
+	// we haven't already contacted.
+	es.peerStatesLk.Lock()
+	for _, p := range lookupRes.peers {
+		if _, found := es.peerStates[p]; found {
+			continue
+		}
+
+		go es.putProviderRecord(p)
+		es.peerStates[p] = Sent
+	}
+	es.peerStatesLk.Unlock()
+
+	// wait until a threshold number of RPCs have completed
+	es.waitForRPCs()
+
+	if outerCtx.Err() == nil && lookupRes.completed { // likely the "completed" field is false but that's not a given
+
+		// tracking lookup results for network size estimator as "completed" is true
+		if err = dht.nsEstimator.Track(key, lookupRes.closest); err != nil {
+			logger.Warnf("network size estimator track peers: %s", err)
+		}
+
+		// refresh the cpl for this key as the query was successful
+		dht.routingTable.ResetCplRefreshedAtForID(kb.ConvertKey(key), time.Now())
+	}
+
+	return outerCtx.Err()
```

Comment:
Looks like TOC-TOU (what if the context is valid in the first if check but not when we return ?) and this does not follow the happy left:
```suggestion
  if err := outerCtx.Err(); err != nil && lookupRes.completed { // likely the "completed" field is false but that's not a given
    return err
  }

	// tracking lookup results for network size estimator as "completed" is true
	if err = dht.nsEstimator.Track(key, lookupRes.closest); err != nil {
		logger.Warnf("network size estimator track peers: %s", err)
	}

	// refresh the cpl for this key as the query was successful
	dht.routingTable.ResetCplRefreshedAtForID(kb.ConvertKey(key), time.Now())
```